### PR TITLE
print reason why parent task was cancelled

### LINKED ISSUE #11830: [Feature Request] we should print detail reason why the parent task cancelled
### Is your feature request related to a problem? Please describe

version: Opensearch2.9

In product, we see many TaskCancelledExceptions as follows, but I can't find why the parent task was cancelled:
```
[2024-01-05T17:06:29,500+0800][ERROR][o.o.b.OpenSearchUncaughtExceptionHandler] [master0] uncaught exception in thread [opensearch[node0][search][T#3]]
org.opensearch.tasks.TaskCancelledException: The parent task was cancelled, shouldn't start any child tasks
        at org.opensearch.tasks.TaskManager$CancellableTaskHolder.registerChildNode(TaskManager.java:666) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.tasks.TaskManager.registerChildNode(TaskManager.java:345) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.action.support.TransportAction.registerChildNode(TransportAction.java:76) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.action.support.TransportAction.execute(TransportAction.java:95) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.client.node.NodeClient.executeLocally(NodeClient.java:110) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.client.node.NodeClient.doExecute(NodeClient.java:97) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.client.support.AbstractClient.execute(AbstractClient.java:476) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.client.support.AbstractClient.search(AbstractClient.java:607) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.action.search.TransportMultiSearchAction.executeSearch(TransportMultiSearchAction.java:180) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.action.search.TransportMultiSearchAction$1.handleResponse(TransportMultiSearchAction.java:203) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.action.search.TransportMultiSearchAction$1.onFailure(TransportMultiSearchAction.java:188) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.action.support.TransportAction$1.onFailure(TransportAction.java:122) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.action.ActionListener$1.onFailure(ActionListener.java:88) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.action.ActionListener$5.onFailure(ActionListener.java:275) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.action.search.AbstractSearchAsyncAction.raisePhaseFailure(AbstractSearchAsyncAction.java:691) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:665) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.action.search.FetchSearchPhase$1.onFailure(FetchSearchPhase.java:128) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:54) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.threadpool.TaskAwareRunnable.doRun(TaskAwareRunnable.java:78) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:59) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:908) ~[opensearch-2.9.0.jar:2.9.0]
        at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) ~[opensearch-2.9.0.jar:2.9.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) [?:?]
```
if we can also print the reason why the parent task cancelled.

### Related component

Search

### PR DESCRIPTION
**Pull request by kkewwei**
### Description
Print reason why parent task was cancelled

### Related Issues
Resolves #11830

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/tasks/TaskManager.java [hunk 1]
```diff
@@ -662,7 +663,7 @@ public CancellableTask getTask() {
 
         synchronized void registerChildNode(DiscoveryNode node) {
             if (banChildren) {
-                throw new TaskCancelledException("The parent task was cancelled, shouldn't start any child tasks");
+                throw new TaskCancelledException("The parent task was cancelled, shouldn't start any child tasks, reason:" + banReason);
```
Comment (dblock) [inline server/src/main/java/org/opensearch/tasks/TaskManager.java:666]: Nitpicking, how about we omit `reason:` and just append `banReason` when present, so `tasks, " + banReason`.

**Review by dblock — COMMENTED** [review]
Is it possible for `banReason` to be blank? If so we should not append `,` or `reason:`.

**Comment by kkewwei** [comment]
@dblock  I have check the banReason can't be null.

`startBan` will be called in `TaskCancellationService.cancelTaskAndDescendants`, it will be called in two place:
1. `BanParentRequestHandler.messageReceived()`: in here, reason in BanParentTaskRequest can't be null.
2. `TaskManager.cancelTaskAndDescendants()`, it also can't be null.

Even if banReason is null, it will not throw NullPointerException.

**Comment by dblock** [comment]
Looks good, let's add a line to CHANGELOG?

**Comment by kkewwei** [comment]
https://build.ci.opensearch.org/blue/organizations/jenkins/gradle-check/detail/gradle-check/42016/pipeline

**Review by dblock — APPROVED** [review]

File: server/src/main/java/org/opensearch/tasks/TaskManager.java [hunk 2]
```diff
@@ -517,10 +517,10 @@ public Set<TaskId> getBannedTaskIds() {
      * @param onChildTasksCompleted called when all child tasks are completed or failed
      * @return the set of current nodes that have outstanding child tasks
      */
-    public Collection<DiscoveryNode> startBanOnChildrenNodes(long taskId, Runnable onChildTasksCompleted) {
+    public Collection<DiscoveryNode> startBanOnChildrenNodes(long taskId, Runnable onChildTasksCompleted, String reason) {
```
[the remark this answers is not in the document]
Comment (kkewwei) [reply server/src/main/java/org/opensearch/tasks/TaskManager.java:520]: maybe some of the plugins depend on the public method, I will add one.

**Review by dblock — APPROVED** [review]

**Pull request merged**